index.js

```diff
@@ -243,12 +276,52 @@ Shopify.prototype.graphql = function graphql(data, variables) {
     parseJson: this.options.parseJson,
     timeout: this.options.timeout,
     responseType: 'json',
-    retry: 0,
     method: 'POST',
-    body: json ? this.options.stringifyJson({ query: data, variables }) : data
+    body: json ? this.options.stringifyJson({ query: data, variables }) : data,
+    retry:
+      this.options.maxRetries > 0
+        ? {
+            limit: this.options.maxRetries,
+            // Don't clamp Shopify retry-after header values too low.
+            maxRetryAfter: Infinity,
+            calculateDelay: (retryObject) => {
+              attemptNumber += 1;
```

Comment:
I think this does not solves the issue. For example assume that a request is answered with 200 but has errors (`request.body.errors`). Right now it does not even increment `attemptNumber` but more importantly even if it does, that counter is not used when the request is answered, for example, with 500. So if the limit is 3 and the request is answered with 200+error, 200+error, 500, 418, the limit might not be respected.

Reply:
> When you call `retryWithUpdatedOptions` in that after response hook, it still comes through this code path, as it is a retry, which I think is smart of got to do that way.

I did not know this. I think it makes sense now. Let me think a bit more about correctness.